peps/pep-0807.rst

```diff
@@ -0,0 +1,398 @@
+  in the context of the interaction (e.g. an internal network).
+
+  Uploading clients **MUST** reject any URLs that do not meet this constraint.
+
+* All server-supplied URLs (i.e. those in discovery responses) **MUST**
+  have the same host subcomponent as the user-provided upload URL. Uploading
+  clients **MUST** reject any URLs that do not meet this constraint.
+
+  In practice, this means that a discovery request to
+  ``https://upload.example.com/.well-known/pytp/{key}`` can only
+  return URLs with the ``upload.example.com`` host.
+
+* All client-made HTTP requests **SHOULD** have an
+  ``Accept: application/vnd.pypi.pytp.v1+json`` header. In the absence of
+  an ``Accept`` header, the receiving server **MUST** behave as if this header
+  were present.
+
+  Receiving servers **SHOULD** respond with a ``406 Not Acceptable``
+  status code if any other ``Accept`` header is present.
+
+.. _discovery:
+
+Trusted Publishing Discovery
+----------------------------
+
+All Python package uploading is currently "endpoint driven," in the sense
+uploading clients (like twine and uv) are given an upload URL (and
+**not** merely a domain name).
+
+For example, to upload to PyPI, uploading clients are expected to connect
+to ``https://upload.pypi.org/legacy/``.
+
+The discovery mechanism proposed below takes advantage of this fact to
+allow single domains to advertise support for multiple indices
+(and their corresponding upload endpoints).
+
+The discovery mechanism is as follows:
+
+1. The uploading client is given an upload URL, e.g.
+   ``https://upload.example.com/legacy/``.
+
+2. The uploading client extracts the *path component* of the URL,
+   as defined in :rfc:`3986`. If the path component is empty,
+   the empty string should be used.
+
+   For the above example, the path component is
+   ``/legacy/``.
+
+3. The uploading client takes the SHA2-256 hash of the path component,
+   producing the *discovery key*.
+
+   For the above example, the discovery key is
+   ``af030c06750716b1b35852298fe852b90def13dcbd012a5fe5148470f1206bfc``. [#fn-hash]_
+
+4. The uploading client constructs a *discovery URL* by taking the
+   scheme and authority components (as defined in :rfc:`3986`)
+   of the upload URL and appending ``/.well-known/pytp/``
+   and the discovery key.
+
+   For the above example, the discovery URL is
+   ``https://upload.example.com/.well-known/pytp/af030c06750716b1b35852298fe852b90def13dcbd012a5fe5148470f1206bfc``.
+
+5. The uploading client performs an HTTP GET request to the discovery URL.
+
+6. The server responds with a ``200 OK`` status code and a body
+   containing a JSON object if the index supports Trusted Publishing
+   for the given upload URL. The JSON object **MUST** contain the following
+   fields:
+
+   - ``audience-endpoint``: a string containing the URL of the OIDC
+     audience endpoint to be used during token exchange.
+   - ``mint-token-endpoint``: a string containing the URL of the
```

Comment:
Perhaps use `token-mint-endpoint` for token minting. Though I suspect that `mint-token-endpoint` matches the currently used name.

Reply:
Yeah, the current endpoint is `mint_token` in Warehouse's codebase, and the current endpoint is `/_/oidc/mint-token`. But this PEP intentionally disconnects those names from the keys in the discovery endpoint, so renaming is fine by me!